### ISSUE TITLE
Drop php support below 5.6?

### ISSUE THREAD

**Issue by burzum** [avalanche123/Imagine#527]
Just see http://php.net/supported-versions.php

**Comment by avalanche123** [comment]
👍

**Comment by robfrawley** [comment]
:+1:

**Comment by romainneutron** [comment]
What value does it bring to drop support? 
Let's think about *the industry*. PHP 5 is still widely used. Encouraging migrating to PHP7 does not mean dropping support of PHP5

As supporting old PHP versions is not a pain at the moment, and updating the dependency, I'm 👎 at the moment

**Comment by burzum** [comment]
@romainneutron php 5.6 is OK but already EOL. It will receive support until 18th December 2018.

I don't see a reason to stick to old, unsupported versions.

**Comment by nicolas-grekas** [comment]
:-1: also, unless you bump the major version number. Having this strict policy is one of the key success factor in Symfony - they know they can keep up to date within the same major version without being forced to change their infra. For sure it adds some burden on the maintainers yet it's really nice to users - and I think every maintainer should and wants to be nice to users, so the burden is worth it.

**Comment by burzum** [comment]
With all the upcoming changes just bump the version? Apply fixes that work for the 0.6.x version and make tests work and tag a new 0.6.x. After that bump to 0.7.0 or even 1.0.0 (if you want to go for semver) and drop php support for versions below 5.6.

**Comment by romainneutron** [comment]
Well @burzum ,

My point is that *the industry* (people working with PHP) are still using PHP 5 a lot. Ubuntu LTS still distribute PHP 5, Symfony LTS are still supporting PHP 5. 
For instance, CakePHP seems to support PHP 5.5 in its latest release too.

Imagine is a high quality library that helps its users, removing PHP 5 support will break imagine usage for lots of users. 
From the experience we have in Symfony, the backward compatibility promise is very important, and such changes should be considered very carefully.

I'd rather be in favor of working towards a Imagine 1.0.0 version and drop PHP 5.3 and 5.4 in a version 2, and wait for PHP 7 wide adoption to drop PHP 5 support in a version 3.

I don't understand your point, I don't understand why you don't want to support PHP 5 anymore. Please let me know.

**Comment by robfrawley** [comment]
PHP 5 has reached "end of life" and that is reason enough for me. It is the responsibility of the community as a whole to help move people along onto newer releases and to avoid the slow move to PHP `5.x` that was facilitated in part by lack of demand from users, due to lack of requirement from developers.

Additionally, the Packagist statistics show that every version of PHP is declining in usage except for PHP 7.x, which was 1% behind surpassing PHP 5.6's usage percentage back in November of 2015. PHP `7.x` is faster for users and offers a better experience for developers. This move _does_ provide quite a few language features.

__For a move from PHP `5.3` to `5.6`:__

- Short array syntax (I'm so tired of using `array()`)
- Variadic functions
- Array unpacking
- `use` function and const
- `finally` blocks in exception handling
- Class name resolution via `::class`
- Support for traits
- Function array dereferencing
- Closures with `$this` context
- Class member access on instantiation (ie `(new Foo)->bar()`)

__For a move all the way to PHP `7.x`:__

- Scalar type declarations
- Return type declarations
- Null coalescing operator
- Spaceship operator
- Array constants using `define`
- Anonymous classes
- Generators
- Generator delegation

Remember, that according to Semvar _"Major version zero (0.y.z) is for initial development. Anything may change at any time. The public API should not be considered stable."_ Additionally, it's not like the current release will somehow disappear. We aren't considering the PHP version bump for the `0.6` series, but instead for our next major release.

The `0.6` series will remain available and I'd recommend we continue to accept all bug fixes and security fixes, as well as feature requests that are reasonable. The new code can (depending on demand) be backported, as well. But in the context of our next major release, I remain :+1: for a big PHP minimum version requirement bump.

**Comment by robfrawley** [comment]
I do believe that we should treat this as a `1.0` release when we bump the PHP requirements, not a `0.7`, regardless of the semvar recommendations.

**Comment by romainneutron** [comment]
@robfrawley IIUC, your point is to drop support of PHP 5 because it EOL'ed. I'm sorry, this is not enough, and you did not answer my arguments. I discussed yours, I would like a dialogue instead of being stucked in front of stubborn arguments.

Do you know Imagine's release history? There's not support for n and n - 1 branches, at the moment, there is only maintenance on the current branch. Maintaining a n - 1 branch for security and bug fix is **huge** work

**Comment by burzum** [comment]
>  I would like a dialogue instead of being stucked in front of stubborn arguments.

Calling other peoples arguments and concerns "stubborn" is not what I would call having a "dialog". It sounds arrogant the best and doesn't add value to the discussion.

I agree with @robfrawley and my suggestion would be to stabilise 0.6, get the tests working again, and fix the bugs that are reported. No new feature additions.

At the same time start working on a 1.0 branch that drops 5.6 support and goes directly to 7. It is not like that the 0.6 branch is unusable. It worked fine for years and even within the last ~2 years where almost no support at all was present. So I doubt anyone will complain if there is a 0.6 version that works for php5 and a modern version that gets new features.

While I agree that software development should be evolution and not revolution I don't see a revolution happen with this proposal because a working and usable solution for people stuck to older php versions will be still available. If "the industry" as you call it wants to stick to old versions fine, but then you have to live with some drawbacks, the same as when you decide to go for bleeding edge versions. I've worked on both kinds of projects before.

You can remove me as a collaborator from the repo by the way, looks like you guys have more time you can and want voluntary spend on it. It's enough for me that it looks like development is going on again now. :)

**Comment by robfrawley** [comment]
@romainneutron I simply expressed my opinion on the matter. Can we entertain the idea of being respectful to others?

**Comment by garak** [comment]
+1 for dropping support for 5.* and go for php7. We need an example from major libraries to take such steps.
Of course, if someone is still on an old php version, they can continue using older versions of this library. Also, if there are major security issues, a new minor release for `0.6` can be released.

By the way: Ubuntu current LTS ships with php7

**Comment by mlocati** [comment]
-1 for dropping support for PHP versions at EOL. Not everybody can easily upgrade their projects.
For instance, we have a server with PHP 5.5 running a huge project. It has been developed during 5 years by 6 people.
Since it's so huge, even if it's full of tests, we can't cover all the possible cases. Furthermore, it's not exposed to the web (it in one of our LANs). So refreshing it's code it's not feasible (I already asked...).

**Comment by garak** [comment]
@mlocati just keep the old code. It will continue to work

**Comment by mlocati** [comment]
I still don't see why we should drop support to old PHP versions...

For instance, I'm looking forward for #585 getting merged: if Imagine drops support for old PHP versions, I'd have to manually apply patches to my installation. All that for no reason.

**Comment by garak** [comment]
Reason is to move along and exploit new features.
We can't stay behind just because you're not able to update your code (but, ehi, you want issues to be closed, so you'll need to update something before of after...)

**Comment by mlocati** [comment]
Which are the features offered by newer PHP versions that are blocking furher development?

PS: furthermore, it seems to me that this package is somehow abandoned...

**Comment by Koc** [comment]
What about increase min php version before 1.0? People who still needed 5.x support can stay on 0.7.x without any troubles.

**Comment by mlocati** [comment]
Why? This library still works very well wih PHP 5.3...

**Comment by Koc** [comment]
For the furure and reasons explained by @robfrawley https://github.com/avalanche123/Imagine/issues/527#issuecomment-278841501

**Comment by garak** [comment]
Meanwhile, current Ubuntu LTS ships with php 7.2

**Comment by mlocati** [comment]
> For the furure and reasons explained by @robfrawley #527 (comment)

I'm fully aware of the benefits of newer PHP versions. BTW Imagine doesn't use *any* of these features, so why should we drop support to it?

**Comment by garak** [comment]
So you can start to use them?

**Comment by mlocati** [comment]
And frankly I'm a bit surprised: the only contributions to this project have been "drop php 5.5- support" (many times), even if I put a lot of efforts in these recent pull requests, where a feedback would have been very welcome...

**Comment by systemist** [comment]
Whoever told @mlocati that upgrading PHP 5.3 is "absolutely out of question" might change their mind once they find out more and more libraries are moving to PHP7.

**Comment by CoalaJoe** [comment]
Would contribution be accepted to move the code base to PHP 7.2? (Current Security Fixes only)

**Comment by mlocati** [comment]
IMHO no. We have a lot of open issues, if you really want to contribute...

**Comment by CoalaJoe** [comment]
I see it this way:

Dropping support for unsupported PHP versions would:
- Reduce code
- Make the code easier to understand

Which in return would make it easier for people to contribute.

**Comment by mlocati** [comment]
I don't think that the reason why people don't contribute to this project is because they have to write `array()` instead of `[]` (or stuff like this).

**Comment by jleckie** [comment]
Just had to write a PR for this library, and having to remember how to write PHP like its 2010 is certainly not straight forward.

**Comment by mdeboer** [comment]
So just because people failed to upgrade to 7.x or 8.x and still use an EOL version is reason enough justification to hold back on the development of this library?

In fact that is the _only_ argument given, contrary to the plenty of arguments given to _do_ upgrade to 7.x.

I am willing to put in some effort to support the Vips library but _only_ if PHP 5.6 support is dropped. Not going back there again.

**Comment by mlocati** [comment]
> So just because people failed to upgrade to 7.x or 8.x and still use an EOL version is reason enough justification to hold back on the development of this library?

What's being hold back?

**Comment by ausi** [comment]
Looking at the [contribution graph](https://github.com/avalanche123/Imagine/graphs/contributors?from=2018-01-01&to=2021-03-11&type=c) I think it’s clear that @mlocati is the maintainer of this library since 2018. If he needs PHP 5 support, we obviously keep it because we don’t want to loose our maintainer.

> I am willing to put in some effort to support the Vips library but _only_ if PHP 5.6 support is dropped. Not going back there again.

I like tabs over spaces, so I could have said that I will not put effort into pull requests until the library switches to tabs. But instead I actually created pull requests for features and bugfixes to improve the library. 

Also, creating a Vips driver can be perfectly done in it’s own library (I did that for an [SVG driver](https://github.com/contao/imagine-svg) for example).

The library is PHP 8.0 compatible with version 1.2.4 (who made the pull request? 😉), nothing is hold back IMO.

So I’d suggest, instead of insisting to drop support for something, start contributing. There is enough work do be done, e.g. review the AVIF/HEIC pull request #759 test it and help me fix the CI issues for example.

**Comment by ninze** [comment]
I generally agree with @ausi and @mlocati that dropping PHP5 support wouldn't magically increase the number of contributions to this project. On the other hand, in order to get more contributors, this project probably needs more users. When I am searching for a library to use, I generally look for an active project with a good community and lots of active developers and modern features that make my coding more enjoyable. I would not pick a php3 or php4 library for my new project. Would I pick php5? Depends... But generally we don't want Imagine to look like old fashioned abandonware and at some point it will be time to let go.

**Comment by ausi** [comment]
My dream solution would be to create a new version 2.0 with support for PHP 7.4+ (or even 8.0+) and keep working on bugfixes (maybe even some features) for 1.x. 

The 2.0 version could then also add support for newer php features like scalar types and more.

But maintaining two versions is a whole lot more work and I can fully understand if the maintainer does not want to go that route. This would most likely also mean more work for contributors having to take two versions into account (for bugfixes at least).

**Comment by garak** [comment]
> But maintaining two versions is a whole lot more work 

That's not true anymore, at least since tools like Rector exist.

**Comment by mdeboer** [comment]
> My dream solution would be to create a new version 2.0 with support for PHP 7.4+ (or even 8.0+) and keep working on bugfixes (maybe even some features) for 1.x.
> 
> The 2.0 version could then also add support for newer php features like scalar types and more.
> 
> But maintaining two versions is a whole lot more work and I can fully understand if the maintainer does not want to go that route. This would most likely also mean more work for contributors having to take two versions into account (for bugfixes at least).

Depends where you draw the line. If you start version 2.0 then you could say 1.x will only receive bug fixes and no new features. Then you're free to fundamentally change 2.0 while people requiring and maintaining the 1.x version can still do so. For all I care they can even propose new features as long as they are ported upstream to 2.0 or 2.0 provides a similar feature.

Tons of projects have moved on, sure this project can do too.

**Comment by mbabker** [comment]
How about using analytical data to justify maintenance burdens?  Composer has a fancy feature showing PHP versions used to install a package, so let's take a look at https://packagist.org/packages/imagine/imagine/php-stats.

PHP 5 accounts for, at best, 3% of all installs of this package since the PHP version stats went live.  If you filter that down to only 1.2 (current minor version branch) installs, the number is basically zero.  If you look at PHP 7.0 and 7.1 users, they still account for maybe 8% of total installs, but the vast majority of those are 0.x installs.

So, there is a lot of argument being made to keep support for EOL PHP versions because someone might need to be able to use the current library features on old PHP versions, but statistically, that user base just does not exist.  The overwhelming majority of library users are on PHP 7.2 and later.  So why justify the maintenance requirement for older PHP versions?

**Comment by ausi** [comment]
> If you filter that down to only 1.2 (current minor version branch) installs, the number is basically zero.

I get around 0.4% for PHP 5 from the stats. With 19,500 installs per day thats about 78 daily installs on PHP 5.

> but statistically, that user base just does not exist.

From my reading of the statistics the library gets installed around 28,470 times on PHP 5 in one year.

**Comment by garak** [comment]
> From my reading of the statistics the library gets installed around 28,470 times on PHP 5 in one year.


What's the point?
Dropping support for old versions will *not* prevent such people from do the same installation another million times

**Comment by mlocati** [comment]
Which issue will be fixed if we drop support dor old php versions?

**Comment by ninze** [comment]
> Which issue will be fixed if we drop support for old php versions?

This very same argument might still be valid in 2030 :)

**Comment by mlocati** [comment]
> This very same argument might still be valid in 2030 :)

I'd ask the very same question even in 2050

**Comment by garak** [comment]
> Which issue will be fixed if we drop support for old php versions?

this very one

**Comment by mbabker** [comment]
> Which issue will be fixed if we drop support for old php versions?

If you are looking for a specific bug report which can only be fixed by dropping legacy PHP support, then my guess is that this is the only one, and either this is an actionable request or should be closed out as a "won't fix" item.

What dropping legacy PHP support does fix, which cannot be measured in a GitHub issue report, is:

- Needing various levels of workarounds to support 10 PHP version branches requiring different combinations of 6 major PHPUnit versions to be able to adequately test, each of those releasing new versions annually
- Needing contributors and maintainers to write code optimized for PHP versions that have been out of support for at least five years
- Needing contributors and maintainers to write code optimized for PHP versions where users aren't even installing the library (see https://packagist.org/packages/imagine/imagine/php-stats/platform/1.2.json for the raw numbers used to generate the stats graph with the current 1.2 version of this library)

**Comment by mlocati** [comment]
@mbabker so, since I'm the de facto maintainer, that's only a problem for me, isn't it?

**Comment by mbabker** [comment]
No.  It's a burden for every potential contributor to this library, as ultimately they too will need to take on those burdens for every possible code change (if writing a feature, they have to ensure it works with PHP 5.3 or has appropriate version guards, they have to write tests compatible with PHPUnit 4.8 or use appropriate version conditionals to make sure tests don't break on different version combinations).  Unless you're basically saying "I don't care how people contribute to this library and they only write code which will execute correctly on PHP 7.4+, I'll do the heavy lifting to deal with version compatibility myself".  Which, while applaudable, I would seriously say isn't worth the effort when you have real data saying that there isn't a real need for a new version of the Imagine library released in 2021 to support PHP 5.3.

**Comment by ausi** [comment]
Just as a side note: none of the open pull requests seem to have an issue with the PHP 5 support: https://github.com/avalanche123/Imagine/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc

**Comment by mbabker** [comment]
Well, then dead serious, if the consensus from those maintaining the library is "we are not changing the minimum supported PHP version", then this issue should be closed.  As long as it remains open, it's just going to keep inviting discussion.

I only chimed in here because I saw the issue open, and thought I could provide new feedback beyond an emotional "old PHP is dead, stop supporting it" standpoint that a lot of people like to lean on by providing real-life usage data to show that supporting old PHP versions is (IMO) an unjustified maintenance burden as the numbers just aren't there for it.  I've done OSS for the last decade and I have no problem justifying legacy PHP version support if there is an actual userbase for it.  But, with the install numbers of this package from Packagist suggesting that the overwhelming majority of Composer users (which also seems to be the only officially supported install mechanism, there is zero documentation on setup and install without it) aren't using Imagine 1.* with PHP 5, I would conclude that even if you're comfortable supporting PHP 5.3 and even if those contributing to this library aren't blocked by the PHP 5.3 requirement, that there is just nobody using the current version of this library and a version of PHP which has been out of support since 2014.

**Comment by PHPGangsta** [comment]
Please drop support of old PHP versions 1 or 2 years after EOL. If you support old PHP versions, you are implicitly telling developers that you are OK with using those old versions. You shouldn't use or support old PHP versions years after EOL, be an example to others.

Libraries like Imagine should encourage users to upgrade their PHP versions every few years. People then still can decide if they stay with an old version of Imagine, or if they upgrade PHP to get the newest features, performance improvements and bugfixes.

A soft push from libraries helps the PHP ecosystem to move forward.
(even Wordpress has 5.6 as Minimum, you shouldn't be behind Wordpress 😄 )

**Comment by mlocati** [comment]
Who am I to force someone to do something?

**Comment by garak** [comment]
> Who am I to force someone to do something?

But, again, by dropping unsupported PHP versions you're not forcing anyone to anything.

**Comment by mlocati** [comment]
> by dropping unsupported PHP versions

By dropping support for old PHP versions we'd force users to either stuck with an old Imagine version or to upgrade their PHP version.

**Comment by mlocati** [comment]
> You can always maintain two branches of a library.

I can barely maintain one...

**Comment by PowerKiKi** [comment]
> I've been told to not touch it

Then what would it change for you if newer releases don't support your old PHP version if you cannot touch your legacy system anyway ?

I feel there's often an irrational fear that "dropping PHP" will make it impossible to use the library with old PHP anymore. Of course that's not true, old versions of the lib will still be perfectly usable.

But more pragmatically, as of today the latest release of this lib is not compatible with PHP 8.1, even though it has been [fixed and merged](https://github.com/php-imagine/Imagine/pull/768) almost 6 months ago. There are a [9 issues](https://github.com/php-imagine/Imagine/milestone/1) that block a new release. And I spent a whole afternoon to create [a PR](https://github.com/php-imagine/Imagine/pull/820) that fixes 7/9 issues. And the way to fix those is to drop support for PHP 5.3 and 5.4, which are now over 10 years old.

So that is one very pragmatic benefit of dropping support for something that is almost no used at all: unblock the support for PHP 8.1.

You asked for what issues would be solved if we dropped *some* PHP versions, here are 7 of them:

- #780
- #781
- #783
- #784
- #785
- #786
- #788

**Comment by PowerKiKi** [comment]
FYI since https://github.com/php-imagine/Imagine/commit/e68f44578ec33d1bbd8d81ea9c0c7b3b6e83be48#diff-d2ab9925cad7eac58e0ff4cc0d251a937ecf49e4b6bf57f8b95aab76648a9d34L20 this project dropped support for PHP 5.3 and 5.4. It's a step in the right direction :tada:
